### PR TITLE
Improving UX around Chat titles

### PR DESCRIPTION
**Pull request by dvargasfuertes**
https://www.loom.com/share/b59403b878f14fa8bb5d9ce7d191bf6c

### CODE REVIEW

File: examples/vercel-ai-chatbot/app/actions.ts [hunk 1]
```diff
@@ -7,11 +7,7 @@ import { kv } from '@vercel/kv'
 import { auth } from '@/auth'
 import { type Chat } from '@/lib/types'
 
-export async function getChats(userId?: string | null) {
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/app/actions.ts:10]: we use typescript, we could always assume it's defined

File: examples/vercel-ai-chatbot/app/actions.ts [hunk 2]
```diff
@@ -22,9 +18,9 @@ export async function getChats(userId?: string | null) {
       pipeline.hgetall(chat)
     }
 
-    const results = await pipeline.exec()
+    const results = await pipeline.exec<Chat[]>()
 
-    return results as Chat[]
+    return results
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/app/actions.ts:23]: we want to discourage "bad" developer patterns wherever possible, like `as` casting and type ignores

File: examples/vercel-ai-chatbot/app/actions.ts [hunk 3]
```diff
@@ -118,7 +163,7 @@ export async function shareChat(id: string) {
     }
   }
 
-  await kv.hmset(`chat:${chat.id}`, chat)
+  await kv.hset(`chat:${chat.id}`, chat)
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/app/actions.ts:169]: `hmset` is deprecated according to [redis](https://redis.io/commands/hmset/)

File: examples/vercel-ai-chatbot/app/api/chat/route.ts [hunk 4]
```diff
@@ -92,19 +92,12 @@ export async function POST(req: Request) {
...
-              id,
-              createdAt: Date.now(),
-              userId,
+            await kv.hset(`chat:${id}`, {
               messages: messages.concat({
                 role: 'ASSISTANT' as const,
                 content: stringOutputType
               })
             })
-            await kv.zadd(`user:chat:${userId}`, {
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/app/api/chat/route.ts:104]: we were adding an entry to redis on every message? comeon Vercel

File: examples/vercel-ai-chatbot/app/api/chat/route.ts [hunk 5]
```diff
@@ -122,10 +115,7 @@ export async function POST(req: Request) {
                   id: nanoid()
                 }) + '\n'
               )
-              await kv.hmset(`chat:${id}`, {
-                id,
-                createdAt: Date.now(),
-                userId,
+              await kv.hset(`chat:${id}`, {
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/app/api/chat/route.ts:118]: now each message invocation will only add the new message

File: examples/vercel-ai-chatbot/components/chat.tsx [hunk 6]
```diff
@@ -5,8 +5,6 @@ import { ChatList } from '@/components/chat-list'
 import { ChatPanel } from '@/components/chat-panel'
 import { EmptyScreen } from '@/components/empty-screen'
 import { ChatScrollAnchor } from '@/components/chat-scroll-anchor'
-import { toast } from 'react-hot-toast'
-import { usePathname, useRouter } from 'next/navigation'
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/components/chat.tsx:9]: unused

File: examples/vercel-ai-chatbot/components/sidebar-item.tsx [hunk 7]
```diff
@@ -22,11 +22,41 @@ interface SidebarItemProps {
...
-  const isActive = pathname === chat.path
+  const isActive = pathname === `/chat/${chat.id}`
   const [newChatId, setNewChatId] = useLocalStorage('newChatId', null)
   const shouldAnimate = index === 0 && isActive && newChatId
 
-  if (!chat?.id) return null
+  const [showContent, setShowContent] = React.useState<boolean>(false)
+  const containerRef = React.useRef<HTMLDivElement>(null)
+
+  React.useEffect(() => {
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/components/sidebar-item.tsx:32]: stole this from `ShowOnHover`

File: examples/vercel-ai-chatbot/lib/hooks/use-vellum-chat.tsx [hunk 8]
```diff
@@ -32,6 +33,20 @@ const useVellumChat = ({
   const triggerRequest = React.useCallback(
     async (request: { messages: ChatMessage[]; id: string }) => {
       setIsLoading(true)
+      if (!path.includes('chat')) {
+        const userMessage = request.messages.find(
+          m => m.role === 'USER' && m.content?.type === 'STRING'
+        )
+        const defaultChatTitle =
+          (userMessage?.content?.value as string)?.slice(0, 10) ?? 'New chat'
+
```
Comment (dvargasfuertes) [inline examples/vercel-ai-chatbot/lib/hooks/use-vellum-chat.tsx:42]: oos but open to ideas - better default title logic

File: examples/vercel-ai-chatbot/app/actions.ts [hunk 9]
```diff
@@ -40,7 +36,56 @@ export async function getChat(id: string, userId: string) {
...
+export async function editChat({ id, title }: { id: string; title: string }) {
+  const session = await auth()
+
+  if (!session) {
+    return {
+      error: 'Unauthorized'
+    }
+  }
+
+  //Convert uid to string for consistent comparison with session.user.id
```
Comment (siddseethepalli) [inline examples/vercel-ai-chatbot/app/actions.ts]:
```suggestion
  // Convert uid to string for consistent comparison with session.user.id
```

File: examples/vercel-ai-chatbot/app/actions.ts [hunk 10]
```diff
@@ -40,7 +36,56 @@ export async function getChat(id: string, userId: string) {
...
+
+  await kv.hset(`chat:${id}`, {
+    title,
+    id,
+    createdAt: Date.now(),
+    userId: session.user.id,
+    messages: []
+  })
+  await kv.zadd(`user:chat:${session.user.id}`, {
+    score: Date.now(),
```
Comment (nihalathanikkal) [inline examples/vercel-ai-chatbot/app/actions.ts:60]: [nit] naming: `score` and `Date.now()` don't seem to match 

If this is unavoidable, might be worth a comment for what score means?
Similar for `member`

File: examples/vercel-ai-chatbot/app/api/chat/route.ts (hunk 4 above)
Comment (siddseethepalli) [reply examples/vercel-ai-chatbot/app/api/chat/route.ts:104]: Is this to set `score` for chat ordering?

File: examples/vercel-ai-chatbot/lib/hooks/use-vellum-chat.tsx (hunk 8 above)
Comment (nihalathanikkal) [reply examples/vercel-ai-chatbot/lib/hooks/use-vellum-chat.tsx:42]: Feels aligned with OpenAI chats which I think is good 👍

**Review by nihalathanikkal — APPROVED** [review]
Amazing 🤩 so excited for this

**Review by siddseethepalli — APPROVED** [review]

File: examples/vercel-ai-chatbot/app/actions.ts (hunk 10 above)
Comment (dvargasfuertes) [reply examples/vercel-ai-chatbot/app/actions.ts:60]: yea these are [redis](https://redis.io/commands/zadd/) specific concepts - I'll add a commend

File: examples/vercel-ai-chatbot/app/api/chat/route.ts (hunk 4 above)
Comment (dvargasfuertes) [reply examples/vercel-ai-chatbot/app/api/chat/route.ts:104]: I suppose to keep the last edited at the top?

**Pull request merged**